---
title: "E484K_210910"
author: "Lily Zhong"
date: "4/24/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Getting results

### Load Packages

```{r load packages, message = FALSE}
# data manipulation
library(tidyverse)
library(stringr)
library(lubridate)
# phylogenetic
library(ape)
library(phytools)
library(castor)
library(phangorn)
# plotting
library(ggplot2)
library(ggExtra)
library(gridExtra)
library(RColorBrewer)
# statistical testing
library(corrplot)
```

### Import data

#### NextStrain Subsample Tree

```{r import NextStrain Subsample Tree}
# read in the tree
tree <- read.tree("nextstrain_timetree_210910.nwk")

# check basics
length(tree$tip.label) # 3449 tree tip labels = 3449 samples
summary(tree$edge.length)
#     Min.    1st Qu.   Median     Mean     3rd Qu.     Max.
# -0.001012  0.026966  0.070278  0.120726  0.166907  1.342243
sum(tree$edge.length < 0) # 9 edges has negative lengths
sum(tree$edge.length == 0) # 257 edges has negative lengths
is.rooted(tree) # TRUE
is.binary(tree) # FALSE: not fully resolved
```

#### NextClade Variant Calling

```{r import NextStrain Subsample NextClade Variant Calling, message = FALSE}
# read in the NextClade csv
sequence <- read_csv2("nextclade_210910.csv")

# check basics
dim(sequence) # (3442, 63) -- missing sequences for 7/3449 of tree tips/samples

# arrange information: 
sequence <- sequence %>% 
            select(seqName, clade, Nextclade_pango, aaSubstitutions, aaDeletions) %>% 
            # merge non-synonymous substitutions with deletions
            mutate(mutation = ifelse(is.na(aaDeletions),
                                     aaSubstitutions, 
                                     paste(aaSubstitutions, aaDeletions, sep = ","))) %>% 
            # assign SARS-CoV-2 variants to sequences
            mutate(variant = case_when(clade %in% c("19A", "19B", "20A", "20B", "20C", "20D", 
                                                    "20E (EU1)", "20F", "20G") ~ "other",
                                       clade == "20I (Alpha, V1)" ~ "Alpha",
                                       clade == "20H (Beta, V2)" ~ "Beta",
                                       clade == "20J (Gamma, V3)" ~ "Gamma",
                                       clade == "21A (Delta)" ~ "Delta",
                                       clade == "21B (Kappa)" ~ "Kappa",
                                       clade == "21C (Epsilon)" ~ "Epsilon",
                                       clade == "21D (Eta)" ~ "Eta",
                                       clade == "21F (Iota)" ~ "Iota",
                                       clade == "21G (Lambda)" ~ "Lambda",
                                       clade == "21H (Mu)" ~ "Mu",
                                       clade == "21I (Delta)" ~ "Delta",
                                       clade == "21J (Delta)" ~ "Delta"))
```

#### Patient Metadata

```{r import NextStrain Subsample Patient Metadata, message = FALSE}
# problem(read_tsv("nextstrain_patient_meta_210910.tsv"))
# format of `Collection date` at line 208 (2021-05-7) and line 2401 (2021-06-1) is wrong
# change to line 208 (2021-05-07) and line 2401 (2021-06-01) in the .tsv file
# read in the metadata tsv
meta <- read_tsv("nextstrain_patient_meta_210910.tsv")

# check basics
dim(meta) # (3442, 17) -- missing sequences for 7/3449 of tree tips/samples
```

### Match sequence names with tree tip names & remove tree tips with no sequence or meta info

```{r Match Sequence Names with Tree Tip Names}
# step 1: extract parts of sequence labels that have the same pattern as the tree tip label
sequence <- sequence %>% 
            mutate(GISAID_ID = seqName) %>% # GISAID_ID for backup
            mutate(seqName = str_extract(.$seqName, "hCoV-19/.+/.+/\\d{4}")) %>% # tree tip
            mutate(Accession_ID = GISAID_ID) %>% 
            mutate(Accession_ID = str_extract(.$Accession_ID, "EPI_ISL_\\d*")) %>% # meta
            select(GISAID_ID, seqName, Accession_ID, variant, mutation)
# check 1
length(intersect(tree$tip.label, sequence$seqName)) # 2910 matches
# find names only in tree tip label
only_tree <- data.frame(tree_lab = setdiff(tree$tip.label, sequence$seqName))
only_tree <- only_tree %>% arrange(tree_lab)
# dim(only_tree) # 539
# find names only in sequence label
only_seq <- sequence %>%
            mutate(seq_lab = seqName) %>%
            select(seq_lab, GISAID_ID) %>%
            # keep GISAID_ID to match back to the sequence data set later
            filter(seq_lab %in% setdiff(sequence$seqName, tree$tip.label)) %>%
            # only in sequence label
            arrange(seq_lab)
# dim(only_seq) # (532, 2)
# add NAs so that the dim matches
only_seq[533:539, ] <- NA
dim(only_seq) # (539, 2)
mismatch <- cbind(only_tree, only_seq)

# step 2: correct for sequence names
mismatch <- mismatch %>%
            # remove spaces in country names
            mutate(seq_lab = str_replace_all(mismatch$seq_lab, "\\s", "")) %>%
            # major changes in region
            mutate(seq_lab = case_when(str_detect(seq_lab, "ed'Iv") ~ str_replace(seq_lab, "'", "-"),
                                       str_detect(seq_lab, "Aruba") ~ str_replace(seq_lab, "Aruba", "Netherlands"),
                                       str_detect(seq_lab, "Bonaire") ~ str_replace(seq_lab, "Bonaire", "Netherlands"),
                                       str_detect(seq_lab, "Curacao") ~ str_replace(seq_lab, "Curacao", "Netherlands"),
                                       str_detect(seq_lab, "FrenchGuiana") ~ str_replace(seq_lab, "FrenchGuiana", "Guyane"),
                                       str_detect(seq_lab, "PuertoRico") ~ str_replace(seq_lab, "PuertoRico", "USA"),
                                       TRUE ~ seq_lab)) %>%
            # minor changes in region
            mutate(seq_lab = case_when(str_detect(seq_lab, "Netherlands/AW-RIVM-51693") ~ str_replace(seq_lab, "Netherlands/AW-RIVM-51693", "Aruba/-RIVM-51693"),
                                       str_detect(seq_lab, "Netherlands/AW-RIVM-52747") ~ str_replace(seq_lab, "Netherlands/AW-RIVM-52747", "Aruba/-RIVM-52747"),
                                       str_detect(seq_lab, "Netherlands/BQ-RIVM-51731") ~ str_replace(seq_lab, "Netherlands/BQ-RIVM-51731", "Bonaire/-RIVM-51731"),
                                       str_detect(seq_lab, "Netherlands/BQ-RIVM-51753") ~ str_replace(seq_lab, "Netherlands/BQ-RIVM-51753", "Bonaire/-RIVM-51753"),
                                       str_detect(seq_lab, "Netherlands/BQ-RIVM-51792") ~ str_replace(seq_lab, "Netherlands/BQ-RIVM-51792", "Bonaire/-RIVM-51792"),
                                       str_detect(seq_lab, "SX-RIVM-52857") ~ str_replace(seq_lab, "SX-RIVM-52857", "-RIVM-52857"),
                                       str_detect(seq_lab, "CzechRepublic/CSQ0207") ~ str_replace(seq_lab, "CzechRepublic/CSQ0207", "Czech_Republic/CSQ0207"),
                                       TRUE ~ seq_lab)) %>%
            # corrections for date: 2021 & 2020
            mutate(seq_lab = case_when(str_detect(seq_lab, "CHRF-0356/2020") ~ str_replace(seq_lab, "CHRF-0356/2020", "CHRF-0356/2021"),
                                       str_detect(seq_lab, "AM-FIOCRUZ-21840050SRD/2020") ~ str_replace(seq_lab, "AM-FIOCRUZ-21840050SRD/2020", "AM-FIOCRUZ-21840050SRD/2021"),
                                       str_detect(seq_lab, "Greece/6462/2020") ~ str_replace(seq_lab, "Greece/6462/2020", "Greece/6462/2021"), 
                                       str_detect(seq_lab, "AGU-InDRE_F11762B_S1084/2020") ~ str_replace(seq_lab, "AGU-InDRE_F11762B_S1084/2020", "AGU-InDRE_F11762B_S1084/2021"),
                                       str_detect(seq_lab, "Oman/520278568/2020") ~ str_replace(seq_lab, "Oman/520278568/2020", "Oman/520278568/2021"),
                                       str_detect(seq_lab, "Oman/5211886/2020") ~ str_replace(seq_lab, "Oman/5211886/2020", "Oman/5211886/2021"),
                                       str_detect(seq_lab, "Oman/52180/2020") ~ str_replace(seq_lab, "Oman/52180/2020", "Oman/52180/2021"),
                                       str_detect(seq_lab, "Romania/SC08965/2020") ~ str_replace(seq_lab, "Romania/SC08965/2020", "Romania/SC08965/2021"), #
                                       TRUE ~ seq_lab)) %>%
            # other minor changes in labels
            mutate(seq_lab = case_when(str_detect(seq_lab, "MG-HLAGYN-1853475/2021") ~ str_replace(seq_lab, "MG-HLAGYN-1853475/2021", "GO-HLAGYN-1853475/2021"),
                                       str_detect(seq_lab, "BKE1545-bc17") ~ str_replace(seq_lab, "BKE1545-bc17", "COV-BKE1545/bc17"),
                                       str_detect(seq_lab, "CRIE-L189B0546b") ~ str_replace(seq_lab, "CRIE-L189B0546b", "UA43-CRIE-L189B0546b"),
                                       str_detect(seq_lab, "COA-InDRE-IBT-28845-NC") ~ str_replace(seq_lab, "COA-InDRE-IBT-28845-NC", "COA-InDRE-IBT-28845/NC"),
                                       str_detect(seq_lab, "DUR-InDRE-IBT-34067-NC") ~ str_replace(seq_lab, "DUR-InDRE-IBT-34067-NC", "DUR-InDRE-IBT-34067/NC"),
                                       str_detect(seq_lab, "VD-GEN2699") ~ str_replace(seq_lab, "VD-GEN2699", "GEN2699"),
                                       str_detect(seq_lab, "Kyiv-F2-6") ~ str_replace(seq_lab, "Kyiv-F2-6", "Kyiv/F2-6"),
                                       str_detect(seq_lab, "GA-CDC-3784545-001") ~ str_replace(seq_lab, "GA-CDC-3784545-001", "GA-CDC-5100"),
                                       str_detect(seq_lab, "IL-CDC-02983522-001") ~ str_replace(seq_lab, "IL-CDC-02983522-001", "IL1"),
                                       str_detect(seq_lab, "NH-CDC-03068248-001") ~ str_replace(seq_lab, "NH-CDC-03068248-001", "NH_0033"),
                                       str_detect(seq_lab, "RI-CDC-03070138-001") ~ str_replace(seq_lab, "RI-CDC-03070138-001", "RI-CDC-KCPZ-3781"),
                                       str_detect(seq_lab, "SD-SDPHL-0361/2020") ~ str_replace(seq_lab, "SD-SDPHL-0361/2020", "SD-SDPHL-0361/2021"),
                                       str_detect(seq_lab, "TN-CDC-3981243-001") ~ str_replace(seq_lab, "TN-CDC-3981243-001", "TN-CDC-TX0A-6344"),
                                       str_detect(seq_lab, "VT-CDC-03068154-001") ~ str_replace(seq_lab, "VT-CDC-03068154-001", "VT-CDC-0303"),
                                       TRUE ~ seq_lab))
            
# check 2: all corresponding ones are matched, the rest are ones with no sequence info
length(intersect(mismatch$seq_lab, mismatch$tree_lab)) # 532 mismatches resolved, 7 to be resolved

# step 3: match the ordering of tree_lab with seq_lab in mismatch data set
mismatch <- full_join(mismatch[1], mismatch[2:3], by = c("tree_lab" = "seq_lab")) %>%
            filter(!is.na(tree_lab))

# step 4: change the sequence names to the matched names & ordering of tree tip labels
# join
length(intersect(tree$tip.label, sequence$seqName)) # 2910/3449 matches
sequence <- full_join(sequence, mismatch, by = "GISAID_ID")
# check
sum(!is.na(sequence$tree_lab)) # 532 matched tree_lab & seq_lab + 7 unpaired tree_labs
# sequence %>% filter(is.na(GISAID_ID)) # 7 missing ones with no sequence info
# match names
sequence <- sequence %>% 
            mutate(seqName = ifelse(!is.na(tree_lab),
                                    # has tree_lab means names were mismatched originally
                                    tree_lab,
                                    # use corrected names from mismatch data set
                                    seqName))
                                    # if not, keep original name
# check again
sum(!is.na(sequence$tree_lab)) # 532 matched tree_lab & seq_lab + 7 unpaired tree_labs
sum(!is.na(sequence$seqName)) # 3449 matched tree_lab & seqName
# sequence %>% filter(is.na(GISAID_ID)) # 7 missing ones with no sequence inf
length(intersect(sequence$seqName, tree$tip.label))
      # 3449/3449 matches between sequence name and tree tip label
setequal(sequence$seqName, tree$tip.label) # TRUE
identical(sequence$seqName, tree$tip.label)# FALSE: the order is wrong

# match order between tree tip label and seqName (sequence)
tree_tip_label <- data.frame(seqName = tree$tip.label) # extract ordering of tree tip label
sequence <- full_join(tree_tip_label, sequence, by = "seqName") # join: correct name order
# check again again
identical(sequence$seqName, tree$tip.label) # TRUE
# sequence %>% filter(is.na(GISAID_ID))
      # 7 missing ones with no sequence inf
# remove intermediate data sets
rm(only_seq); rm(only_tree); rm(mismatch); rm(tree_tip_label)
``` 

```{r remove tips with no squence or meta info then merge sequence and meta}
# extract tree tip labels with missing sequence information
noSeq <- sequence$seqName[which(is.na(sequence$GISAID_ID))]
# check tree parameters before removing tips with no sequence information
length(tree$tip.label) # 3449 tips
tree$Nnode # 2827 internal nodes
length(tree$edge.length) # 6275 edges
dim(tree$edge) # (6275, 2)
# remove the tips from the sequence data set
tree <- drop.tip(tree, noSeq, trim.internal = TRUE)
      # Phylogenetic tree with 3442 tips and 2801 internal nodes.
      # Tip labels:
      # hCoV-19/Wuhan/Hu-1/2019, hCoV-19/Wuhan/WH01/2019, hCoV-19/Taiwan/2/2020, 
      # hCoV-19/HongKong/HKPU23_2601/2020, hCoV-19/Thailand/SI200383-NT/2020, 
      # hCoV-19/Australia/VIC03/2020, ...
      # Rooted; includes branch lengths.
# check tree parameters after removing tips with no sequence information
length(tree$tip.label) # 3442 tips (-7 tips)
tree$Nnode # 2801 internal nodes (-26 internal nodes)
length(tree$edge.length) # 6242 edges (-33 edges)
dim(tree$edge) # (6242, 2)
summary(tree$edge.length)
#     Min.    1st Qu.   Median     Mean     3rd Qu.     Max.
# -0.001012  0.027213  0.070801  0.121154  0.167211  1.342243 
sum(tree$edge.length < 0) # 9 edges has negative lengths
sum(tree$edge.length == 0) # 237 edges has negative lengths

# merge sequence with meta & clean the data set for further processing 
sequence <- sequence %>%
            filter(!is.na(GISAID_ID)) %>%
            full_join(meta, by = c("Accession_ID" = "Accession ID")) %>% 
            select(GISAID_ID, seqName, `Virus name`, Accession_ID, variant, 
                   mutation, `Collection date`, Location)
```

### E484 Ancestral State Reconstruction

#### Create numeric vector for E484 status of each sequence

```{r E484 vectors for mutations}
E484 <- sequence$mutation # E484 now has all mutations in each sequence
# sum(is.na(E484)) # 8 NAs = 8 sequences with 0 mutations
E484 <- ifelse(is.na(E484), "E", E484) # no mutation = "E" on amino acid number 484
# sum(is.na(E484)) # check: 0 NA
# sum(str_detect(E484, "(S:E484)([A-Z])")) # 494 sequences have substitution (any amino acid) on site 484
E484 <- ifelse(str_detect(E484, "(S:E484)([A-Z])"), 
               str_match(E484, "(S:E484)([A-Z])")[, 3], # if mutation, extract the substitution
               "E") # if no "E484" mutation pattern found = "E" on site 484
# sum(E484 != "E") # check: 494 matches
rbind(table(E484), prop.table(table(E484))) # calculate substitution patterns on E484

# create numeric vector of tip mutation for ASR
E484 <- data.frame(E484)
names(E484) <- "AA"
    # E484A, E484K, E484Q
E484 <- E484 %>% mutate(indAA = case_when(AA == "A" ~ 1,
                                          AA == "E" ~ 2, 
                                          AA == "K" ~ 3,
                                          AA == "Q" ~ 4))
E484_int <- as.integer(E484$indAA) # E484 for backup, E484_int for ASR: 1, 2, 3, 4
rbind(table(E484_int), prop.table(table(E484_int))) # check

# assigned matched names to E484_int for ASR
# identical(sequence$seqName, tree$tip.label) # TRUE: order and labels are the same
names(E484_int) <- sequence$seqName
identical(names(E484_int), tree$tip.label) # TRUE
```

#### E484 Ancestral State Reconstruction

```{r E484 Ancestral State Reconstruction}
# ASR with castor::asr_mk_model()
E484_ASR_mk <- asr_mk_model(tree, E484_int, Nstates = 4, rate_model = "ER", 
                            include_ancestral_likelihoods = TRUE, check_input = TRUE)
# dim(E484_ASR_mk$ancestral_likelihoods) # 2801 nodes * 4 states
E484_likelihood_mk <- E484_ASR_mk$ancestral_likelihoods
# from probabilities to states for internal nodes
E484_states <- c("A", "E", "K", "Q")
E484_mk <- NaN
for (i in 1:dim(E484_likelihood_mk)[1]){
  E484_mk[i] = E484_states[which.max(E484_likelihood_mk[i, ])]
  # max probability = state of internal nodes
}
table(E484_mk, useNA = "ifany") # E(2369) K(421) Q(11) = state count of internal nodes
```

#### E484 Clade Identification

```{r identify clades based on E484 ASR}
# data frame: store the most-recent common ancestor information for each tip
clade <- data.frame(tip_label = names(E484_int),
                    # tip labels in format "hCoV-19/AREA/SEQUENCE/YEAR"
                    tip_node = 1:Ntip(tree), # node number of tips
                    tip_state = E484$AA, # "E, K, Q, R" state of tree tip
                    ancestor_node = NaN, # will hold ancestor node
                    ancestor_state = NaN, # will hold ASR ancestor node
                    row.names = NULL)

# vector of G204 state on all tips and internal nodes (tips followed by nodes)
E484_tip_and_node <- c(E484$AA, E484_mk)

# find the oldest ancestor (deepest node with the same state) for each tip
for (tip in clade$tip_node){ 
  # will hold the node number of the oldest ancestor with the same mutation
  find = NaN 
  # loop over and check all ancestors (not just parent) of a tip
  for (ancestor in Ancestors(tree, tip, type = "all")){ 
    # if ancestor node state from ASR = tip state
    if (E484_mk[ancestor-Ntip(tree)] == E484$AA[tip]){ 
      # update ancestor if the state of the ancestor node is the same as the tip
      find = ancestor 
    } else{break}
  }
  # if no ancestor matches, use the tip itself
  find <- ifelse(is.na(find), tip, find)
  # match the ancestor and the corresponding tip
  clade$ancestor_node[which(clade$tip_node == tip)] <- find 
  # fill in the ASR state of the ancestor
  clade$ancestor_state[which(clade$tip_node == tip)] <- ifelse(find == tip, # if tip itself
                                                               E484$AA[find], # tip state
                                                               E484_mk[find-Ntip(tree)])
                                                               # ancestor state
}

# check: should be the same
identical(clade$tip_state, clade$ancestor_state) # TRUE
```

### Characterize E484 ASR clades

#### Size, duration, earliest sample, lastest sample of a clade

```{r E484 clade characteristics}
# find size of each clade defined by number of descending tips
clade <- clade %>% 
         group_by(ancestor_node) %>%
         mutate(size = n()) %>% 
         ungroup()

# number of emergence of each state and related clade information
emerge <- clade %>% 
          # filter(ancestor_node > Ntip(tree)) %>% # exclude tips, keep only internal nodes
          select(ancestor_state, ancestor_node, size) %>%
          unique() %>%
          group_by(ancestor_state) %>%
          mutate(times = n()) %>% # number of emergence for each state
          ungroup() %>%
          arrange(ancestor_state) %>% 
          mutate(tips = NaN) %>%
          select(ancestor_state, times, ancestor_node, size, tips) # reorder the columns

# store and join the descending tips information
for (anc in emerge$ancestor_node) {
  emerge$tips[which(emerge$ancestor_node == anc)] <- clade %>% 
                                                     filter(ancestor_node == anc) %>%
                                                     select(tip_node) %>% unique() %>%
                                                     .$tip_node %>% toString()
}

# reorder: most frequently emerging substitution -> largest clade size
emerge <- emerge %>% arrange(desc(times), desc(size))
clade <- clade %>% 
         select(ancestor_state, ancestor_node, size, tip_node, tip_label) %>%
         arrange(ancestor_state, desc(size))

# merge clade with sequence via tip_label=seqName, keep Accession_ID and mutation
clade <- clade %>% left_join(sequence, by = c("tip_label" = "seqName"))

# find the earliest and latest sample collection date of each clade
clade <- clade %>%
         group_by(ancestor_node) %>% 
         mutate(earliest = min(`Collection date`, na.rm = TRUE)) %>%
         mutate(latest = max(`Collection date`, na.rm = TRUE)) %>%
         mutate(duration = as.numeric(latest - earliest)) %>%
         # how long a clade lasted in day
         ungroup() %>% 
         mutate(tip_date_day = `Collection date` - ymd("2019-12-26")) %>%
         # day since earliest
         mutate(tip_date_week = time_length(tip_date_day, "weeks")) %>%
         # week since earliest
         mutate(tip_date_month = time_length(tip_date_day, "months")) %>%
         # month since earliest
         mutate(tip_date_day = as.numeric(tip_date_day)) 

# add time information to emerge
temp <- clade %>%
        group_by(ancestor_node) %>%
        select(ancestor_node, earliest, latest, duration) %>%
        unique() %>%
        ungroup()
emerge <- emerge %>%
          full_join(temp, by = "ancestor_node")
rm(temp)
```

#### E484 Reference and Mutated Clades

```{r}
# clades with no E484 mutation
E_clade <- clade %>% filter(ancestor_state == "E")
E_emerge <- emerge %>% filter(ancestor_state == "E")

# clades with any E484 mutation
E484_clade <- clade %>% filter(ancestor_state != "E")
E484_emerge <- emerge %>% filter(ancestor_state != "E")

# area of emergence of E484-mutated clades
# function to extract continent information from Location
continent <- function(x) {  
  str_split(x, "/") [[1]][1] %>% trimws(which = "both")
}
E484_clade$Continent <- sapply(E484_clade$Location, continent)

# function to extract country information from Location
country <- function(x) {  
  str_split(x, "/") [[1]][2] %>% trimws(which = "both")
}
E484_clade$Country <- sapply(E484_clade$Location, country)
```

### Standardized E484K Emergence

```{r standardized distribution of emergence over time}
# find E484K tips that are not the first sample of a E484K clade
later_K <- E484_clade %>% 
           filter(ancestor_state == "K") %>%
           group_by(ancestor_node) %>%
           filter(`Collection date` != earliest) %>%
           ungroup() %>%
           .$tip_node

# all other nodes and tips (E, E484K, E484A, E484Q)
# excluding E484K tips that are not the first sample of a E484K clade
clade_first_K <- clade %>%
                 filter(!(tip_node %in% later_K)) %>%
                 mutate(ind_K = relevel(as.factor(ifelse(ancestor_state == "K", 1, 0)),
                                        ref = "0")) # indicator for E484K mutation (0)

# create categorical time for each two-week period
# summary(clade_first_K$tip_date_week) # range = (0, 88.14)
clade_first_K$tip_date_biweek <- cut(clade_first_K$tip_date_week,
                                     breaks=c(seq(0, 90, 2)),
                                     labels=c(seq(1, 45)))
clade_first_K$tip_date_biweek[which(clade_first_K$tip_date_day == 0)] <- 1
```

### Genetic Background of E484 Clades (Proportion of Other Mutations)

#### Proportion in All E484K Sequences

```{r genetic background at the time of E484 mutation emergence}
# total number of sequences with E484K
K_total <- E484_clade %>% filter(ancestor_state == "K") %>% dim() %>% .[1] # 478
# E484_emerge %>% filter(ancestor_state == "K") %>% .$size %>% sum() # confirm 478 = K_total

# calculate the count and proportion of other mutations in all E484K sequences
other <- E484_clade %>%
         filter(ancestor_state == "K") %>%
         select(mutation) %>% mutate(single = str_split(mutation, ",")) %>% 
         unnest(single) %>% select(single) %>% 
         group_by(single) %>% mutate(count = n()) %>% unique() %>% ungroup() %>%
         arrange(desc(count)) %>% 
         mutate(proportion = count/K_total) %>%
         filter(single != "S:E484K")

# find potential epistatic mutations as: >= 10% percentage among all E484K sequences
key_mut <- other %>% filter(proportion >= 0.1)
```

#### Proportion in Each E484 Clade

```{r}
# define a function to calculate the proportion of specified mutations in each E484K clade
clade_prop <- function(ancestor){
                      # calculate the size of clade descending from an ancestor node
                      size <- emerge$size[which(emerge$ancestor_node == ancestor)]
                      # calculate the proportion of other mutations in this clade
                      prop <- clade %>% 
                              filter(ancestor_node == ancestor) %>%
                              select(mutation) %>% 
                              mutate(single = str_split(mutation, ",")) %>% 
                              unnest(single) %>% select(single) %>% 
                              group_by(single) %>% mutate(clade_count = n()) %>%
                              unique() %>% ungroup() %>%
                              arrange(desc(clade_count)) %>% 
                              mutate(clade_proportion = clade_count/size)
                      # find the proportion of potential_mutation in the clade
                      prop <- prop %>% 
                              full_join(key_mut[, "single"], by = "single") %>%
                              mutate_all(~replace(., is.na(.), 0)) %>%
                              # 0 if no such mutation 
                              filter(single %in% key_mut$single)
                      prop <- key_mut[, "single"] %>% 
                              full_join(prop, by = "single") %>% # reorder
                              select(single, clade_proportion) # remove clade_count
                      # ensure different names for the proportion column of different clades
                      names(prop) <- c("single", paste0("clade_", ancestor))
                      return(prop)
}

# find all ancestor nodes of E484K clade, excluding clades with only 1 tip (size==1)
ancestor_K <- E484_emerge %>% filter(ancestor_state == "K" & size > 1) %>% .$ancestor_node
# find all ancestor nodes of E484Q clade, excluding clades with only 1 tip (size==1)
ancestor_other <- E484_emerge %>% filter(ancestor_state != "K" & size > 1) %>% .$ancestor_node
# find all ancestor nodes of E484E ref clade, excluding clades with only 1 tip (size==1)
ancestor_E <- E_emerge %>% filter(size > 1) %>% .$ancestor_node

# calculate proportion of key_mut in all E484K clades (size > 1)
key_mut_K <- key_mut
for (anc in ancestor_K){
  key_mut_K <- key_mut_K %>% full_join(clade_prop(anc), by = "single")
}
# calculate proportion of key_mut in all E484Q/R clades (size > 1)
key_mut_other <- key_mut
for (anc in ancestor_other){
  key_mut_other <- key_mut_other %>% full_join(clade_prop(anc), by = "single")
}
key_mut_other <- key_mut_other %>% .[ , c(1, 4)] # keep necessary info to merge to key_mut_K
# E484_emerge %>% filter(ancestor_node == 5280) # E484Q-mutated clade
names(key_mut_other) <- c("single", "clade_5280_Q")

# calculate proportion of key_mut in all E484E clades (size > 1)
key_mut_E <- key_mut
for (anc in ancestor_E){
  key_mut_E <- key_mut_E %>% full_join(clade_prop(anc), by = "single")
}
key_mut_E <- key_mut_E %>% .[ , c(1, 4, 5)] # keep necessary info to merge to key_mut_K
names(key_mut_E) <- c("single", "clade_3443_E", "clade_3691_E")
```

### Epistasis with S:N501Y, S:R203K, N:G204R

#### S:N501

```{r N501 vectors for mutations}
N501 <- sequence$mutation # N501 now has all mutations in each sequence
# sum(is.na(N501)) # 8 NAs = 8 sequences with 0 mutations
N501 <- ifelse(is.na(N501), "N", N501) # no mutation = "N" on amino acid number 501
# sum(is.na(N501)) # check: 0 NA
# sum(str_detect(N501, "(S:N501)([A-Z])")) # 896 sequences have substitution (any amino acid) on site 501
N501 <- ifelse(str_detect(N501, "(S:N501)([A-Z])"), 
               str_match(N501, "(S:N501)([A-Z])")[, 3], # if mutation, extract the substitution
               "N") # if no "N501" mutation pattern found = "N" on site 501
# sum(N501 != "N") # check: 701 matches
rbind(table(N501), prop.table(table(N501))) # calculate substitution patterns on N501

# create numeric vector of tip mutation for ASR
N501 <- data.frame(N501)
names(N501) <- "AA"
      # N501S, N501T, N501Y
N501 <- N501 %>% mutate(indAA = case_when(AA == "N" ~ 1,
                                          AA == "S" ~ 2, 
                                          AA == "T" ~ 3,
                                          AA == "Y" ~ 4))
N501_int <- as.integer(N501$indAA) # N501 for backup, N501_int for ASR: 1, 2, 3, 4
rbind(table(N501_int), prop.table(table(N501_int))) # check

# assigned matched names to N501_int for ASR
# identical(sequence$seqName, tree$tip.label) # TRUE: order and labels are the same
names(N501_int) <- sequence$seqName
identical(names(N501_int), tree$tip.label) # TRUE
```

```{r N501 Ancestral State Reconstruction}
# ASR with castor::asr_mk_model()
N501_ASR_mk <- asr_mk_model(tree, N501_int, Nstates = 4, rate_model = "ER", 
                            include_ancestral_likelihoods = TRUE, check_input = TRUE)
# dim(N501_ASR_mk$ancestral_likelihoods) # 2801 nodes * 4 states
N501_likelihood_mk <- N501_ASR_mk$ancestral_likelihoods
# from probabilities to states for internal nodes
N501_states <- c("N", "S", "T", "Y")
N501_mk <- NaN
for (i in 1:dim(N501_likelihood_mk)[1]){
  N501_mk[i] = N501_states[which.max(N501_likelihood_mk[i, ])]
  # max probability = state of internal nodes
}
table(N501_mk, useNA = "ifany") # N(2015) T(4) Y(782) = state count of internal nodes
```

```{r identify clades based on N501 ASR}
# data frame: store the most-recent common ancestor information for each tip
clade_N501 <- data.frame(tip_label = names(N501_int),
                         # tip labels in format "hCoV-19/AREA/SEQUENCE/YEAR"
                         tip_node = 1:Ntip(tree), # node number of tips
                         tip_state = N501$AA, # "E, K, Q, R" state of tree tip
                         ancestor_node = NaN, # will hold ancestor node
                         ancestor_state = NaN, # will hold ASR ancestor node
                         row.names = NULL)

# vector of N501 state on all tips and internal nodes (tips followed by nodes)
N501_tip_and_node <- c(N501$AA, N501_mk)

# find the oldest ancestor (deepest node with the same state) for each tip
for (tip in clade_N501$tip_node){ 
  # will hold the node number of the oldest ancestor with the same mutation
  find = NaN 
  # loop over and check all ancestors (not just parent) of a tip
  for (ancestor in Ancestors(tree, tip, type = "all")){ 
    # if ancestor node state from ASR = tip state
    if (N501_tip_and_node[ancestor] == N501_tip_and_node[tip]){ 
      # update ancestor if the state of the ancestor node is the same as the tip
      find = ancestor 
    } else{break}
  }
  # if no ancestor matches, use the tip itself
  find <- ifelse(is.na(find), tip, find)
  # match the ancestor and the corresponding tip
  clade_N501$ancestor_node[which(clade_N501$tip_node == tip)] <- find 
  # fill in the ASR state of the ancestor
  clade_N501$ancestor_state[which(clade_N501$tip_node == tip)] <- N501_tip_and_node[find]
}

# check: should be the same
identical(clade_N501$tip_state, clade_N501$ancestor_state) # TRUE
```

```{r N501 clade characteristics}
# size of each clade approximated by number of descending tips
clade_N501 <- clade_N501 %>% 
         group_by(ancestor_node) %>%
         mutate(size = n()) %>% 
         ungroup()

# number of emergence of each state and related clade information
emerge_N501 <- clade_N501 %>% 
          # filter(ancestor_node > Ntip(tree)) %>% # exclude tips, keep only internal nodes
          select(ancestor_state, ancestor_node, size) %>%
          unique() %>%
          group_by(ancestor_state) %>%
          mutate(times = n()) %>% # number of emergence for each state
          ungroup() %>%
          arrange(ancestor_state) %>% 
          mutate(tips = NaN) %>%
          select(ancestor_state, times, ancestor_node, size, tips) # reorder the columns

# store and join the descending tips information
for (anc in emerge_N501$ancestor_node) {
  emerge_N501$tips[which(emerge_N501$ancestor_node == anc)] <- clade_N501 %>% 
                                                     filter(ancestor_node == anc) %>%
                                                     select(tip_node) %>% unique() %>%
                                                     .$tip_node %>% toString()
}

# reorder: most frequently emerging substitution -> largest clade size
emerge_N501 <- emerge_N501 %>% arrange(desc(times), desc(size))
clade_N501 <- clade_N501 %>% 
         select(ancestor_state, ancestor_node, size, tip_node, tip_label) %>%
         arrange(ancestor_state, desc(size))

# merge clade with sequence via tip_label=seqName, keep Accession_ID and mutation
clade_N501 <- clade_N501 %>% left_join(sequence, by = c("tip_label" = "seqName"))
```

#### N:R203K

```{r R203 vectors for mutations}
R203 <- sequence$mutation # R203 now has all mutations in each sequence
# sum(is.na(R203)) # 8 NAs = 8 sequences with 0 mutations
R203 <- ifelse(is.na(R203), "R", R203) # no mutation = "R" on amino acid number 203
# sum(is.na(R203)) # check: 0 NA
sum(str_detect(R203, "(N:R203)([A-Z]|-)")) # 2538 sequences have substitution (any amino acid) on site 203
R203 <- ifelse(str_detect(R203, "(N:R203)([A-Z]|-)"), 
               str_match(R203, "(N:R203)([A-Z]|-)")[, 3], # if mutation, extract the substitution
               "R") # if no "R203" mutation pattern found = "R" on site 203
# sum(R203 != "R") # check: 2538 matches
rbind(table(R203), prop.table(table(R203))) # calculate substitution patterns on R203

# create numeric vector of tip mutation for ASR
R203 <- data.frame(R203)
names(R203) <- "AA"
# R203-, R203I, R203K, R203M, R203S
R203 <- R203 %>% mutate(indAA = case_when(AA == "-" ~ 1,
                                          AA == "I" ~ 2, 
                                          AA == "K" ~ 3,
                                          AA == "M" ~ 4,
                                          AA == "R" ~ 5,
                                          AA == "S" ~ 6))
R203_int <- as.integer(R203$indAA) # R203 for backup, R203_int for ASR: 1, 2, 3, 4, 5, 6
rbind(table(R203_int), prop.table(table(R203_int))) # check

# assigned matched names to R203_int for ASR
# identical(sequence$seqName, tree$tip.label) # TRUE: order and labels are the same
names(R203_int) <- sequence$seqName
identical(names(R203_int), tree$tip.label) # TRUE
```

```{r R203 Ancestral State Reconstruction}
# ASR with castor::asr_mk_model()
R203_ASR_mk <- asr_mk_model(tree, R203_int, Nstates = 6, rate_model = "ER", 
                            include_ancestral_likelihoods = TRUE, check_input = TRUE)
# dim(R203_ASR_mk$ancestral_likelihoods) # 2801 nodes * 6 states
R203_likelihood_mk <- R203_ASR_mk$ancestral_likelihoods
# from probabilities to states for internal nodes
R203_states <- c("-", "I", "K", "M", "R", "S")
R203_mk <- NaN
for (i in 1:dim(R203_likelihood_mk)[1]){
  R203_mk[i] = R203_states[which.max(R203_likelihood_mk[i, ])]
  # max probability = state of internal nodes
}
table(R203_mk, useNA = "ifany") # K(1095) M(966) R(740) = state count of internal nodes
```

```{r identify clades based on R203 ASR}
# data frame: store the most-recent common ancestor information for each tip
clade_R203 <- data.frame(tip_label = names(R203_int),
                         # tip labels in format "hCoV-19/AREA/SEQUENCE/YEAR"
                         tip_node = 1:Ntip(tree), # node number of tips
                         tip_state = R203$AA, # "G, R, L, P, -" state of tree tip
                         ancestor_node = NaN, # will hold ancestor node
                         ancestor_state = NaN, # will hold ASR ancestor node
                         row.names = NULL)

# vector of R203 state on all tips and internal nodes (tips followed by nodes)
R203_tip_and_node <- c(R203$AA, R203_mk)

# find the oldest ancestor (deepest node with the same state) for each tip
for (tip in clade_R203$tip_node){ 
  # will hold the node number of the oldest ancestor with the same mutation
  find = NaN 
  # loop over and check all ancestors (not just parent) of a tip
  for (ancestor in Ancestors(tree, tip, type = "all")){ 
    # if ancestor node state from ASR = tip state
    if (R203_tip_and_node[ancestor] == R203_tip_and_node[tip]){ 
      # update ancestor if the state of the ancestor node is the same as the tip
      find = ancestor 
    } else{break}
  }
  # if no ancestor matches, use the tip itself
  find <- ifelse(is.na(find), tip, find)
  # match the ancestor and the corresponding tip
  clade_R203$ancestor_node[which(clade_R203$tip_node == tip)] <- find 
  # fill in the ASR state of the ancestor
  clade_R203$ancestor_state[which(clade_R203$tip_node == tip)] <- R203_tip_and_node[find]
}

# check: should be the same
identical(clade_R203$tip_state, clade_R203$ancestor_state) # TRUE
```

```{r R203 clade characteristics}
# size of each clade approximated by number of descending tips
clade_R203 <- clade_R203 %>% 
         group_by(ancestor_node) %>%
         mutate(size = n()) %>% 
         ungroup()

# number of emergence of each state and related clade information
emerge_R203 <- clade_R203 %>% 
          # filter(ancestor_node > Ntip(tree)) %>% # exclude tips, keep only internal nodes
          select(ancestor_state, ancestor_node, size) %>%
          unique() %>%
          group_by(ancestor_state) %>%
          mutate(times = n()) %>% # number of emergence for each state
          ungroup() %>%
          arrange(ancestor_state) %>% 
          mutate(tips = NaN) %>%
          select(ancestor_state, times, ancestor_node, size, tips) # reorder the columns

# store and join the descending tips information
for (anc in emerge_R203$ancestor_node) {
  emerge_R203$tips[which(emerge_R203$ancestor_node == anc)] <- clade_R203 %>% 
                                                     filter(ancestor_node == anc) %>%
                                                     select(tip_node) %>% unique() %>%
                                                     .$tip_node %>% toString()
}

# reorder: most frequently emerging substitution -> largest clade size
emerge_R203 <- emerge_R203 %>% arrange(desc(times), desc(size))
clade_R203 <- clade_R203 %>% 
         select(ancestor_state, ancestor_node, size, tip_node, tip_label) %>%
         arrange(ancestor_state, desc(size))

# merge clade with sequence via tip_label=seqName, keep Accession_ID and mutation
clade_R203 <- clade_R203 %>% left_join(sequence, by = c("tip_label" = "seqName"))
```

#### N:G204R

```{r G204R vectors for mutations}
G204 <- sequence$mutation # G204 now has all mutations in each sequence
# sum(is.na(G204)) # 8 NAs = 8 sequences with 0 mutations
G204 <- ifelse(is.na(G204), "G", G204) # no mutation = "G" on amino acid number 204
# sum(is.na(G204)) # check: 0 NA
sum(str_detect(G204, "(N:G204)([A-Z]|-)")) # 1276 sequences have substitution (any amino acid) on site 204
G204 <- ifelse(str_detect(G204, "(N:G204)([A-Z]|-)"), 
               str_match(G204, "(N:G204)([A-Z]|-)")[, 3], # if mutation, extract the substitution
               "G") # if no "G204" mutation pattern found = "G" on site 204
# sum(G204 != "G") # check: 1276 matches
rbind(table(G204), prop.table(table(G204))) # calculate substitution patterns on G204

# create numeric vector of tip mutation for ASR
G204 <- data.frame(G204)
names(G204) <- "AA"
# G204-, G204L, G204P, G204R
G204 <- G204 %>% mutate(indAA = case_when(AA == "-" ~ 1,
                                          AA == "G" ~ 2, 
                                          AA == "L" ~ 3,
                                          AA == "P" ~ 4,
                                          AA == "R" ~ 5))
G204_int <- as.integer(G204$indAA) # G204 for backup, G204_int for ASR: 1, 2, 3, 4, 5
rbind(table(G204_int), prop.table(table(G204_int))) # check

# assigned matched names to G204_int for ASR
# identical(sequence$seqName, tree$tip.label) # TRUE: order and labels are the same
names(G204_int) <- sequence$seqName
identical(names(G204_int), tree$tip.label) # TRUE
```

```{r G204 Ancestral State Reconstruction}
# ASR with castor::asr_mk_model()
G204_ASR_mk <- asr_mk_model(tree, G204_int, Nstates = 5, rate_model = "ER", 
                            include_ancestral_likelihoods = TRUE, check_input = TRUE)
# dim(G204_ASR_mk$ancestral_likelihoods) # 2801 nodes * 5 states
G204_likelihood_mk <- G204_ASR_mk$ancestral_likelihoods
# from probabilities to states for internal nodes
G204_states <- c("-", "G", "L", "P", "R")
G204_mk <- NaN
for (i in 1:dim(G204_likelihood_mk)[1]){
  G204_mk[i] = G204_states[which.max(G204_likelihood_mk[i, ])]
  # max probability = state of internal nodes
}
table(G204_mk, useNA = "ifany") # G(1708) L(1) P(32) R(1060) = state count of internal nodes
```

```{r identify clades based on G204 ASR}
# data frame: store the most-recent common ancestor information for each tip
clade_G204 <- data.frame(tip_label = names(G204_int),
                         # tip labels in format "hCoV-19/AREA/SEQUENCE/YEAR"
                         tip_node = 1:Ntip(tree), # node number of tips
                         tip_state = G204$AA, # "G, R, L, P, -" state of tree tip
                         ancestor_node = NaN, # will hold ancestor node
                         ancestor_state = NaN, # will hold ASR ancestor node
                         row.names = NULL)

# vector of G204 state on all tips and internal nodes (tips followed by nodes)
G204_tip_and_node <- c(G204$AA, G204_mk)

# find the oldest ancestor (deepest node with the same state) for each tip
for (tip in clade_G204$tip_node){ 
  # will hold the node number of the oldest ancestor with the same mutation
  find = NaN 
  # loop over and check all ancestors (not just parent) of a tip
  for (ancestor in Ancestors(tree, tip, type = "all")){ 
    # if ancestor node state from ASR = tip state
    if (G204_tip_and_node[ancestor] == G204_tip_and_node[tip]){ 
      # update ancestor if the state of the ancestor node is the same as the tip
      find = ancestor 
    } else{break}
  }
  # if no ancestor matches, use the tip itself
  find <- ifelse(is.na(find), tip, find)
  # match the ancestor and the corresponding tip
  clade_G204$ancestor_node[which(clade_G204$tip_node == tip)] <- find 
  # fill in the ASR state of the ancestor
  clade_G204$ancestor_state[which(clade_G204$tip_node == tip)] <- G204_tip_and_node[find]
}

# check: should be the same
identical(clade_G204$tip_state, clade_G204$ancestor_state) # TRUE
```

```{r G204 clade characteristics}
# size of each clade approximated by number of descending tips
clade_G204 <- clade_G204 %>% 
         group_by(ancestor_node) %>%
         mutate(size = n()) %>% 
         ungroup()

# number of emergence of each state and related clade information
emerge_G204 <- clade_G204 %>% 
          # filter(ancestor_node > Ntip(tree)) %>% # exclude tips, keep only internal nodes
          select(ancestor_state, ancestor_node, size) %>%
          unique() %>%
          group_by(ancestor_state) %>%
          mutate(times = n()) %>% # number of emergence for each state
          ungroup() %>%
          arrange(ancestor_state) %>% 
          mutate(tips = NaN) %>%
          select(ancestor_state, times, ancestor_node, size, tips) # reorder the columns

# store and join the descending tips information
for (anc in emerge_G204$ancestor_node) {
  emerge_G204$tips[which(emerge_G204$ancestor_node == anc)] <- clade_G204 %>% 
                                                     filter(ancestor_node == anc) %>%
                                                     select(tip_node) %>% unique() %>%
                                                     .$tip_node %>% toString()
}

# reorder: most frequently emerging substitution -> largest clade size
emerge_G204 <- emerge_G204 %>% arrange(desc(times), desc(size))
clade_G204 <- clade_G204 %>% 
         select(ancestor_state, ancestor_node, size, tip_node, tip_label) %>%
         arrange(ancestor_state, desc(size))

# merge clade with sequence via tip_label=seqName, keep Accession_ID and mutation
clade_G204 <- clade_G204 %>% left_join(sequence, by = c("tip_label" = "seqName"))
```


## Obtaining/ Plotting Results for Thesis

#### Characterize time distribution of samples

```{r}
# find the time span of the NextStrain subsample
summary(sequence$`Collection date`)
#         Min.      1st Qu.       Median         Mean      3rd Qu.         Max. 
# "2019-12-26" "2021-01-21" "2021-05-28" "2021-03-26" "2021-07-09" "2021-09-03"
```

#### Table 1: distribution of E484 mutations in samples and in variants

```{r}
# proportion of E484 mutations in the NextStrain subsample
rbind(table(E484$AA), round(prop.table(table(E484$AA))*100, 1))

# find the proportion of each variants in each variant
basic <- cbind(c("20I", "20H", "21A, 21I, 21J", "21C", "21D", "20J", "21F", "21B", 
                 "21G", "21H", "other"),
               table(sequence$variant),
               round(prop.table(table(sequence$variant))*100, 1),
               table(sequence$variant, E484$AA)[, 3],
               round(prop.table(table(sequence$variant, E484$AA), margin = 1)*100, 1)[, 3]) %>%
         data.frame()
colnames(basic) <- c("Nextstrain Clade", "Sample Count", "Sample Proportion", 
                     "E484K Count", "E484K Proportion")

basic

# output table to a csv file
# write.csv(basic, file = "basic.csv", row.names = TRUE)

# geographical distribution of E484K - Continent
E484_clade %>% filter(ancestor_state == "K") %>% .$Continent %>% table() %>% prop.table() * 100
E484_clade %>% filter(ancestor_state == "Q") %>% .$Continent %>% table() %>% prop.table() * 100
E484_clade %>% filter(ancestor_state == "A") %>% .$Continent %>% table() %>% prop.table() * 100

# geographical distribution of E484K - Country
E484_clade %>% filter(ancestor_state == "K") %>% .$Country %>% table() %>% prop.table() * 100
E484_clade %>% filter(ancestor_state == "Q") %>% .$Country %>% table() %>% prop.table() * 100
E484_clade %>% filter(ancestor_state == "A") %>% .$Country %>% table() %>% prop.table() * 100
```

#### Basic statistics of E484 clades (size > 1)

```{r}
# size summary of E484-mutated clades (size >= 2)
summary(E484_emerge[which(E484_emerge$size >= 2), ]$size)
#    Min.   1st Qu. Median   Mean  3rd Qu.  Max.
#    2.00    2.75   10.00   39.92   40.75  209.00

# duration summary of E484-mutated clades (size >= 2)
summary(E484_emerge[which(E484_emerge$size >= 2), ]$duration)
#    Min.   1st Qu. Median   Mean  3rd Qu.  Max. 
#    5.00   43.25  112.50  130.75  192.75  303.00 

# geographical distribution of sequences for each E484 clade - Continent
for (loc in E484_emerge$ancestor_node[which(E484_emerge$size > 1)]){
  tmp <- E484_clade %>% filter(ancestor_node == loc)
  print(paste0(loc, " ", E484_emerge$ancestor_state[(which(E484_emerge$ancestor_node == loc))]))
  print(cbind(table(tmp$Continent), prop.table(table(tmp$Continent)) * 100))
}

# geographical distribution of sequences for each E484 clade - Country
for (loc in E484_emerge$ancestor_node[which(E484_emerge$size > 1)]){
  tmp <- E484_clade %>% filter(ancestor_node == loc)
  print(paste0(loc, " ", E484_emerge$ancestor_state[(which(E484_emerge$ancestor_node == loc))]))
  print(cbind(table(tmp$Country), prop.table(table(tmp$Country)) * 100))
}
```

#### Basic statistics of E484 singletons (size == 1)

```{r}
# count of E484-mutated emergence singletons (size == 1)
sum(E484_emerge$size == 1 & E484_emerge$ancestor_state == "K") # 10

# variant of sequences for each E484 clade - Continent
E484_clade %>% filter(ancestor_state == "K") %>% filter(size == 1) %>% .$variant %>% table()

# geographical distribution of sequences for each E484 clade - Continent
for (loc in E484_emerge$ancestor_node[which(E484_emerge$size == 1)]){
  tmp <- E484_clade %>% filter(ancestor_node == loc)
  print(paste0(loc, " ", E484_emerge$ancestor_state[(which(E484_emerge$ancestor_node == loc))]))
  print(cbind(table(tmp$Continent), prop.table(table(tmp$Continent)) * 100))
}

# geographical distribution of sequences for each E484 clade - Country
for (loc in E484_emerge$ancestor_node[which(E484_emerge$size == 1)]){
  tmp <- E484_clade %>% filter(ancestor_node == loc)
  print(paste0(loc, " ", E484_emerge$ancestor_state[(which(E484_emerge$ancestor_node == loc))]))
  print(cbind(table(tmp$Country), prop.table(table(tmp$Country)) * 100))
}
```

#### Figure 4: time distribution of samples in E484 clades

```{r}
# Figure 4 - plot time distribution of E484 clades
E484_clade %>% 
  arrange(ancestor_state, size) %>%
  ggplot(aes(x = `Collection date`, y = as.factor(ancestor_node))) +
  geom_point(aes(color = as.factor(ancestor_state)), size = 3) +
  scale_colour_manual(name = "E484\nMutation",
                      values = c("blueviolet", "brown3", "goldenrod2")) +
  scale_x_date(breaks = c(date("2020-10-01"), date("2020-12-01"),
                          date("2021-02-01"), date("2021-04-01"), 
                          date("2021-06-01"), date("2021-08-01")),
               date_labels = "%Y %b") + 
  labs(title = "Time Distribution of E484 Mutation Emergence", 
       x = "Collection Date", y = "Ancestor Node") +
  theme(text = element_text(size = 24),
        axis.text.x = element_text(size = 18, hjust = 0),
        axis.text.y = element_text(size = 18),
        legend.text = element_text(size = 18))
```

#### (Not in paper: Size and Duration of clade vs. Time of clade emergence)

```{r Figure 5 to be, eval = FALSE}
# Figure 5a to be - plot size of clade vs. duration of clade
sVd <- E484_emerge %>% 
  ggplot(aes(x = duration, y = size)) +
  geom_point(aes(color = as.factor(ancestor_state)), size = 6) +
  scale_colour_manual(name = "E484\nMutation",
                      values = c("blueviolet", "brown3", "goldenrod2")) +
  labs(title = "E484-mutated Clade Size vs. Clade Duration", 
       x = "Clade Duration (day)" , y = "Clade Size") +
  theme(text = element_text(size = 24),
        axis.text.x = element_text(size = 18, hjust = 0),
        axis.text.y = element_text(size = 18),
        legend.position = "none")
ggMarginal(sVd, type = "histogram", fill = "lightblue4", color = "grey50", size = 7)

# Figure 5b to be - plot duration of clade vs. time of earliest sample
E484_emerge %>% 
  ggplot(aes(x = earliest, y = duration)) +
  geom_point(aes(size = size, color = as.factor(ancestor_state))) +
  scale_size_area(name = "Clade\nSize", max_size = 12) +
  scale_colour_manual(name = "E484\nMutation",
                      values = c("blueviolet", "brown3", "goldenrod2")) +
  scale_x_date(breaks = c(date("2020-10-01"), date("2020-12-01"),
                          date("2021-02-01"), date("2021-04-01"), 
                          date("2021-06-01"), date("2021-08-01")),
               date_labels = "%Y %b") +
  labs(title = "E484-mutated Clade Duration vs. Time of Earliest Sample", 
       x = "Time of Emergence" , y = "Clade Duration (day)") +
  theme(text = element_text(size = 24),
        axis.text.x = element_text(size = 18, hjust = 0),
        axis.text.y = element_text(size = 18),
        legend.text = element_text(size = 18))

# calculate the Spearman correlation between size, duration and time
cor.test(E484_emerge$duration, as.numeric(E484_emerge$earliest - as.Date("2019-12-26")), 
         method = "spearman", exact = FALSE) # duration & time, ties
cor.test(E484_emerge$size, as.numeric(E484_emerge$earliest - as.Date("2019-12-26")), 
         method = "spearman", exact = FALSE) # size & time, ties
cor.test(E484_emerge$size, E484_emerge$duration, 
         method = "spearman", exact = FALSE) # size & duration, ties
```

#### Figure 5: standardized frequencey of E484K emergence

```{r}
# Figure 5. plot standardized emergence by time (fornights since first sample)
clade_first_K %>% 
  filter(ancestor_state %in% c("E", "K")) %>% # remove E484A and E484Q tips (already mutated)
#  filter(!ancestor_node %in% c(3712, 3681)) %>% # sensitivity analysis: merge 3712 & 3681 to 3611
  group_by(as.factor(tip_date_biweek)) %>%
  mutate(biweek_size = n()) %>%
  mutate(biweek_emerge = sum(ind_K == 1)) %>%
  mutate(biweek_prop = biweek_emerge / biweek_size) %>%
  ungroup() %>%
  ggplot(aes(as.numeric(tip_date_biweek), biweek_prop)) +
  # ref line: Pfizer BioNTech vaccine authorized
  geom_vline(aes(xintercept = time_length(ymd("2020-12-11") - ymd("2019-12-26"), "weeks")/2,
             color = "Vaccine\nAuthorized"), size = 2.5) +
  geom_smooth(color = "coral3", size = 2.5) +
  geom_point(color = "brown3", size = 5) +
  scale_x_continuous(limits = c(0, 48), breaks = seq(0, 48, 4),
                     labels = c("2019 Dec 26", "2020 Feb 20", "2020 Apr 16", "2020 Jun 11",
                                "2020 Aug 06", "2020 Oct 01", "2020 Nov 26", "2021 Jan 21",
                                "2021 Mar 18", "2021 May 13", "2021 Jul 08", "2021 Sep 02",
                                "2021 Oct 28")) +
  scale_y_continuous(limits = c(0, 0.1)) +
  scale_color_manual(values = c("Vaccine\nAuthorized" = "bisque4"), name = "")  +
  labs(x = "Time of Emergence (Two Weeks)",
       y = "Standardized Frequency of Emergence",
       title = "Standardized Emergence by Time") +
  theme(text = element_text(size = 24),
        axis.text.x = element_text(size = 18, angle = 30, hjust = 1),
        axis.text.y = element_text(size = 18))
```

#### Figure 6: heatmap for genetic background of E484 clades

```{r}
# find which variant each clade belongs to
heatmap.label <- c(E484_emerge[which(E484_emerge$size > 1), ]$ancestor_node, 
                   E_emerge[which(E_emerge$size > 1), ]$ancestor_node)
for (hmp in heatmap.label){
  print(hmp); clade %>% filter(ancestor_node == hmp) %>% .$variant %>% table() %>% print()
}
heatmap.label <- c("Gamma\nNode 4399", "Beta/Mu\nNode 3611", "Eta\nNode 3571", "Node 4123", 
                   "Iota\nNode 3975", "Node 4211", "Node 4271", "Node 4102", "Beta\nNode 3681",
                   "Beta\nNode 3712", "Alpha\nNode 4879", "Kappa\nNode 5280", "*\nNode 3443",
                   "Beta\nNode 3691")

# plot heat map to identify potentially interesting mutations
key_mut_K %>%
  full_join(key_mut_other, by = "single") %>%
  full_join(key_mut_E, by = "single") %>%
  pivot_longer(cols = starts_with("clade_"), 
               names_to = "clade", 
               values_to = "clade_proportion") %>%
  mutate(percent = clade_proportion * 100) %>%
  ggplot(aes(fct_inorder(as.factor(clade)), fct_rev(fct_inorder(as.factor(single))))) +
  geom_tile(aes(fill = clade_proportion), color = "grey") +
  scale_fill_gradientn(name = "Proportion", colors = brewer.pal(9, "Reds"), na.value = "grey") +
  scale_x_discrete(position = "top", labels = heatmap.label) +
  labs(x = NULL, y = NULL, title = NULL) +
  theme(text = element_text(size = 14)) +
  theme(legend.position="bottom", legend.margin = margin(t = -8),
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 12, vjust = 0.8))
```

#### Aesthetics for plotting phylogenetic trees
```{r}
# a data frame for all tips and nodes, with E484, N501, G204, R203 status
tip_and_node <- cbind(1:(length(tree$tip.label) + tree$Nnode),
                      E484_tip_and_node, N501_tip_and_node,
                      G204_tip_and_node, R203_tip_and_node) %>% 
                data.frame()
names(tip_and_node) <- c("tip_node", names(tip_and_node)[-1])
tip_and_node$tip_node <- as.numeric(tip_and_node$tip_node)

summary(as.factor(tip_and_node$E484_tip_and_node))
summary(as.factor(tip_and_node$N501_tip_and_node))
summary(as.factor(tip_and_node$G204_tip_and_node))
summary(as.factor(tip_and_node$R203_tip_and_node))

# data set of all edges: time (edge length), E484 & (S:N501, N:G204R, N:R203K) states
edge_all <- data.frame(tree$edge)
colnames(edge_all) <- c("start", "end")
# add edge length and fix edge length
edge_all <- edge_all %>%
            mutate(time = tree$edge.length) %>%
            mutate(time = case_when(time < 0 ~ abs(time), # (time<0) -> absolute values
                                    time == 0 ~ 0.0000000001, # (time=0) -> 0.0000000001
                                    TRUE ~ time)) %>%
            mutate(time = 365 * time) # change unit from year to day
# add mutation info for the start and end nodes of the edges
edge_all <- edge_all %>% 
            mutate(E484_start = E484_tip_and_node[start]) %>% 
            mutate(E484_end = E484_tip_and_node[end]) %>%
            mutate(N501_start = N501_tip_and_node[start]) %>%
            mutate(N501_end = N501_tip_and_node[end]) %>%
            mutate(G204_start = G204_tip_and_node[start]) %>% 
            mutate(G204_end = G204_tip_and_node[end]) %>%
            mutate(R203_start = R203_tip_and_node[start]) %>%
            mutate(R203_end = R203_tip_and_node[end])

# define the color of tips and nodes by E484 or N501 substitution
tip_and_node <- tip_and_node %>% 
                mutate(color_E484 = case_when(E484_tip_and_node == "A" ~ "blueviolet",
                                              E484_tip_and_node == "E" ~ "black",
                                              E484_tip_and_node == "K" ~ "brown3",
                                              E484_tip_and_node == "Q" ~ "goldenrod2")) %>%
                mutate(tip_E484 = case_when(E484_tip_and_node == "A" ~ "blueviolet",
                                              E484_tip_and_node == "E" ~ "gray50",
                                              E484_tip_and_node == "K" ~ "brown3",
                                              E484_tip_and_node == "Q" ~ "goldenrod2")) %>%
                mutate(color_N501 = case_when(N501_tip_and_node == "N" ~ "gray50",
                                              N501_tip_and_node == "S" ~ "darkorchid",
                                              N501_tip_and_node == "T" ~ "darkolivegreen4",
                                              N501_tip_and_node == "Y" ~ "steelblue2")) %>%
                mutate(color_G204 = case_when(G204_tip_and_node == "-" ~ "beige",
                                              G204_tip_and_node == "G" ~ "gray50",
                                              G204_tip_and_node == "L" ~ "darkorchid",
                                              G204_tip_and_node == "P" ~ "darkolivegreen4",
                                              G204_tip_and_node == "R" ~ "steelblue2")) %>%
                mutate(color_R203 = case_when(R203_tip_and_node == "-" ~ "beige",
                                              R203_tip_and_node == "I" ~ "darkorchid",
                                              R203_tip_and_node == "K" ~ "steelblue2",
                                              R203_tip_and_node == "M" ~ "darkolivegreen4",
                                              R203_tip_and_node == "R" ~ "gray50",
                                              R203_tip_and_node == "S" ~ "aquamarine4"))

# define the color of edges by E484 or N501 clade
edge_all <- edge_all %>%
            mutate(edge_color_E484K = case_when((E484_start == "K" & E484_end == "K") ~
                                                  "brown3",
                                                (E484_start == "Q" & E484_end == "Q") ~ 
                                                  "goldenrod2",
                                                TRUE ~ "gray50")) %>%
            mutate(edge_color_N501Y = case_when((N501_start == "Y" & N501_end == "Y") ~
                                                  "steelblue2",
                                                (N501_start == "T" & N501_end == "T") ~ 
                                                  "darkolivegreen4",
                                                TRUE ~ "gray50")) %>%
            mutate(edge_color_G204R = case_when((G204_start == "L" & G204_end == "L") ~
                                                  "darkorchid",
                                                (G204_start == "P" & G204_end == "P") ~ 
                                                  "darkolivegreen4",
                                                (G204_start == "R" & G204_end == "R") ~ 
                                                  "steelblue2",
                                                TRUE ~ "gray50")) %>%
            mutate(edge_color_R203K = case_when((R203_start == "K" & R203_end == "K") ~
                                                  "steelblue2",
                                                (R203_start == "M" & R203_end == "M") ~ 
                                                  "darkolivegreen4",
                                                TRUE ~ "gray50"))

# E484-mutated tips
E484_mut_tip <- which(E484_tip_and_node[1:length(tree$tip.label)] != "E")
```

#### Figure 2

```{r}
# plot phylogenetic tree colored by E484 clades
plot(tree, type = "phylogram", direction = "right", show.tip.label = FALSE,
     edge.color = edge_all$edge_color_E484K,
     tip.color = tip_and_node$color_E484[1:3442])
tiplabels(tip = c(1:3442), 
          col = tip_and_node$tip_E484[1:3442], pch = 20, cex = 0.5)
nodelabels(node = emerge$ancestor_node[-1], 
           col = tip_and_node$color_E484[emerge$ancestor_node[-1]], 
           pch = 4, cex =3)
# nodelabels(node = c(E484_emerge$ancestor_node, 3691), 
#            col = tip_and_node$color_E484[c(E484_emerge$ancestor_node, 3691)], 
#            pch = 4, cex = 0.8,
#            text = c(E484_emerge$ancestor_node, 3691), frame = "none")
legend("topleft", inset = c(0.16, 0), cex = 0.8, bty = "n",
       col = c("black", "blueviolet", "brown3", "goldenrod2"), 
       pch = c(16, 16), pt.cex = 1.2, 
       legend = c("E", "E484A", "E484K", "E484Q"))
legend("topleft", inset = c(0.26, 0), cex = 0.8, bty = "n",
       col = "black", pch = 4, pt.cex = 1.2,
       legend = "Emergence")
legend("topleft", inset = c(0, 0), cex = 0.8, bty = "n",
       col = c("gray50", "brown3", "goldenrod2"), lty = c(1, 1), lwd = 2, 
       legend = c("E", "E484K", "E484Q"))
```

#### Figure 7

```{r}
# plot phylogenetic tree colored by N501 clades
plot(tree, type = "phylogram", direction = "right", show.tip.label = FALSE,
     edge.color = edge_all$edge_color_N501Y,
     tip.color = tip_and_node$color_E484[1:3442])
tiplabels(tip = c(1:3442), 
          col = tip_and_node$tip_E484[1:3442], pch = 20, cex = 0.5)
nodelabels(node = emerge$ancestor_node[-1], 
           col = tip_and_node$color_E484[emerge$ancestor_node[-1]], 
           pch = 4, cex =3)
nodelabels(node = unique(clade_N501$ancestor_node)[-1], 
           col = tip_and_node$color_N501[unique(clade_N501$ancestor_node)[-1]], 
           pch = 3, cex = 2)
legend("topleft", inset = c(0.16, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "blueviolet", "brown3", "goldenrod2"), 
       pch = c(4, rep(16, 4)), pt.cex = 1.2, 
       legend = c("E484 Emergence", "E", "E484A", "E484K", "E484Q"))
legend("topleft", inset = c(0.31, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "darkorchid", "darkolivegreen4", "steelblue2"), 
       pch = c(3, rep(16, 4)), pt.cex = 1.2,
       legend = c("N501 Emergence", "N", "N501S", "N501T", "N501Y"))
legend("topleft", inset = c(0, 0), cex = 0.8, bty = "n",
       col = c("gray50", "steelblue2"), lty = c(1, 1), lwd = 2, 
       legend = c("N", "N501Y"))
```

#### odds ratio of E484K on nodes with other mutations

```{r}
# numerator is number of emergence (nodes & tips) = E484K
# denominator is number of nodes that are N501Y & N501 & E484K clade ancestral nodes

# for the denominator, find and exclude the set of nodes & tips that are either:
# (1) E484A or E484Q = 1 + 26 = 27 nodes & tips
exclude_E484 <- tip_and_node %>% filter(E484_tip_and_node %in% c("A", "Q")) %>% .$tip_node
# (2) descendants of E484K emergence (not the ancestor of an E484K clade)
# for each E484K ancestral node, find its descendant that's also E484K
for (ancestor in ancestor_K){
  for (descendant in Descendants(tree, ancestor, type = "all")){
    if (tip_and_node$E484_tip_and_node[descendant] == "K"){
      exclude_E484 <- c(exclude_E484, descendant)
    }
  }
}
exclude_E484 <- setdiff(exclude_E484, 
                        E484_emerge$ancestor_node[which(E484_emerge$ancestor_state == "K")])
length(exclude_E484) # 905 nodes excluded; denominator only has E484 and E484K ancestral node

# N501Y & N501 background
N501.node.fisher <- tip_and_node %>% 
                    filter(!(tip_node %in% exclude_E484)) %>%
                    filter(N501_tip_and_node %in% c("N", "Y")) %>% 
                    # further exclude nodes & tips not N501 or N501Y
                    select(E484_tip_and_node, N501_tip_and_node) %>%
                    table() # 5328 nodes & tips left

# G204R & G204 background
G204.node.fisher <- tip_and_node %>% 
                    filter(!(tip_node %in% exclude_E484)) %>%
                    filter(G204_tip_and_node %in% c("G", "R")) %>% 
                    # further exclude nodes & tips not G204 or G204R
                    select(E484_tip_and_node, G204_tip_and_node) %>%
                    table() # 5265 nodes & tips left

# R203K & R203 background
R203.node.fisher <- tip_and_node %>% 
                    filter(!(tip_node %in% exclude_E484)) %>%
                    filter(R203_tip_and_node %in% c("R", "K")) %>% 
                    # further exclude nodes & tips not N501 or N501Y
                    select(E484_tip_and_node, R203_tip_and_node) %>%
                    table() # 3138 nodes & tips left

# fisher's exact test
#               unexposed   exposed
# ref (484E)        a          b
# mut (484-sub)     c          d
fisher.test(N501.node.fisher)
fisher.test(G204.node.fisher)
fisher.test(R203.node.fisher)
```

#### Figure S1: phylogenetic tree with edges colored by R203 clades

```{r}
# plot phylogenetic tree colored by R203 clades
plot(tree, type = "phylogram", direction = "right", show.tip.label = FALSE,
     edge.color = edge_all$edge_color_R203K,
     tip.color = tip_and_node$color_E484[1:3442])
tiplabels(tip = c(1:3442), 
          col = tip_and_node$tip_E484[1:3442], pch = 20, cex = 0.5)
nodelabels(node = emerge$ancestor_node[-1], 
           col = tip_and_node$color_E484[emerge$ancestor_node[-1]], 
           pch = 4, cex =3)
nodelabels(node = unique(clade_R203$ancestor_node)[-14], 
           col = tip_and_node$color_R203[unique(clade_R203$ancestor_node)[-14]], 
           pch = 3, cex = 2)
legend("topleft", inset = c(0.16, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "blueviolet", "brown3", "goldenrod2"), 
       pch = c(4, rep(16, 4)), pt.cex = 1.2, 
       legend = c("E484 Emergence", "E", "E484A", "E484K", "E484Q"))
legend("topleft", inset = c(0.31, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "beige", "darkorchid", "steelblue2", "darkolivegreen4", "aquamarine4"), 
       pch = c(3, rep(16, 6)), pt.cex = 1.2,
       legend = c("R203 Emergence", "R", "R203-", "R203I", "R203K", "R203M", "R203S"))
legend("topleft", inset = c(0, 0), cex = 0.8, bty = "n",
       col = c("gray50", "steelblue2", "darkolivegreen4"), 
       lty = c(1, 1, 1), lwd = 2, 
       legend = c("R", "R203K", "R203M"))
```

#### Figure S2: phylogenetic tree with edges colored by G204 clades

```{r}
# plot phylogenetic tree colored by G204 clades
plot(tree, type = "phylogram", direction = "right", show.tip.label = FALSE,
     edge.color = edge_all$edge_color_G204R,
     tip.color = tip_and_node$color_E484[1:3442])
tiplabels(tip = c(1:3442), 
          col = tip_and_node$tip_E484[1:3442], pch = 20, cex = 0.5)
nodelabels(node = emerge$ancestor_node[-1], 
           col = tip_and_node$color_E484[emerge$ancestor_node[-1]], 
           pch = 4, cex =3)
nodelabels(node = unique(clade_G204$ancestor_node)[-2], 
           col = tip_and_node$color_G204[unique(clade_G204$ancestor_node)[-2]], 
           pch = 3, cex = 2)
legend("topleft", inset = c(0.16, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "blueviolet", "brown3", "goldenrod2"), 
       pch = c(4, rep(16, 4)), pt.cex = 1.2, 
       legend = c("E484 Emergence", "E", "E484A", "E484K", "E484Q"))
legend("topleft", inset = c(0.31, 0), cex = 0.8, bty = "n",
       col = c("black", "gray50", "beige", "darkorchid", "darkolivegreen4", "steelblue2"), 
       pch = c(4, rep(16, 5)), pt.cex = 1.2,
       legend = c("G204 Emergence", "G", "G204-", "G204L", "G204P", "G204R"))
legend("topleft", inset = c(0, 0), cex = 0.8, bty = "n",
       col = c("gray50", "darkorchid", "darkolivegreen4", "steelblue2"), 
       lty = c(1, 1), lwd = 2, 
       legend = c("G", "G204L", "G204P", "G204R"))
```

#### Figure S3a: correlation matrix of E484 clades (by within-clade proportion of other mutations)

```{r}
# only proportions, no names
key_mut_matrix <- key_mut_K %>%
                  full_join(key_mut_other, by = "single") %>%
                  full_join(key_mut_E, by = "single") %>%
                  select(-c(single, count, proportion)) %>%
                  data.frame()
rownames(key_mut_matrix) <- key_mut_K$single
colnames(key_mut_matrix) <- heatmap.label

# correlation matrix
corrplot(cor(key_mut_matrix), method = "color", order = "original",
         tl.cex = 1.2, cl.cex = 1.2, tl.col="black")
```

#### Figure S3b: sub-phylogenetic tree from node 3611 (clades 3712 and 3681 embedded)

```{r}
# extract subtrees from node 3611
subtree_3611 <- extract.clade(tree, 3611)
subtree_3611
length(subtree_3611$tip.label) + subtree_3611$Nnode # 276
# plot(subtree_3611)

# match tree and sub-tree
sub_tip_and_node <- data.frame("tip_node" = c(3611, Descendants(tree, 3611, type = "all")),
                    "sub_tip_node" = c(150, Descendants(subtree_3611, 150, type = "all"))) %>% 
                    arrange(tip_node)
sub_tip_and_node <- inner_join(sub_tip_and_node, tip_and_node, by = "tip_node")
sub_edge_all <- data.frame(subtree_3611$edge)
colnames(sub_edge_all) <- c("sub_start", "sub_end")
sub_edge_all <- sub_edge_all %>%
                mutate(E484_start = sub_tip_and_node$E484_tip_and_node[sub_start]) %>%
                mutate(E484_end = sub_tip_and_node$E484_tip_and_node[sub_end]) %>%
                mutate(edge_color_E484K = case_when((E484_start == "K" & E484_end == "K") ~
                                                     "brown3",
                                                    (E484_start == "Q" & E484_end == "Q") ~ 
                                                     "goldenrod2",
                                                    TRUE ~ "gray50"))

# plot sub-phylogenetic tree 3611 colored by E484 clades
plot(subtree_3611, type = "phylogram", direction = "right", show.tip.label = FALSE,
     edge.color = sub_edge_all$edge_color_E484K,
     tip.color = sub_tip_and_node$tip_E484[1:149])
tiplabels(tip = c(1:149), 
          col = sub_tip_and_node$tip_E484[1:149], pch = 20, cex = 0.5)
nodelabels(node = sub_tip_and_node$sub_tip_node[which(sub_tip_and_node$tip_node %in%
                                                        emerge$ancestor_node)], 
           col = sub_tip_and_node$color_E484[sub_tip_and_node$sub_tip_node[which(sub_tip_and_node$tip_node %in% emerge$ancestor_node)]], 
           pch = 4, cex =3)
nodelabels(node = c(150, 251, 220), 
           col = "brown3", 
           pch = 4, cex = 0.8,
           text = c(3611, 3712, 3681), frame = "none")
legend("topleft", inset = c(0.16, 0), cex = 0.8, bty = "n",
       col = c("black", "blueviolet", "brown3", "goldenrod2"), 
       pch = c(16, 16), pt.cex = 1.2, 
       legend = c("E", "E484A", "E484K", "E484Q"))
legend("topleft", inset = c(0.26, 0), cex = 0.8, bty = "n",
       col = "black", pch = 4, pt.cex = 1.2,
       legend = "Emergence")
legend("topleft", inset = c(0, 0), cex = 0.8, bty = "n",
       col = c("gray50", "brown3", "goldenrod2"), lty = c(1, 1), lwd = 2, 
       legend = c("E", "E484K", "E484Q"))
```